test/runtest.py

```diff
@@ -805,5 +805,239 @@ def remove_temporary_files():
 
     ray.worker.cleanup()
 
+class ResourcesTest(unittest.TestCase):
+
+  def testResourceConstraints(self):
+    num_workers = 20
+    ray.init(num_workers=num_workers, num_cpus=10, num_gpus=2)
+
+    # Attempt to wait for all of the workers to start up.
+    ray.worker.global_worker.run_function_on_all_workers(lambda worker_info: sys.path.append(worker_info["counter"]))
+    @ray.remote(num_cpus=0)
+    def get_worker_id():
+      time.sleep(1)
+      return sys.path[-1]
+    while True:
+      if len(set(ray.get([get_worker_id.remote() for _ in range(num_workers)]))) == num_workers:
+        break
+
+    time_buffer = 0.3
+
+    # At most 10 copies of this can run at once.
+    @ray.remote(num_cpus=1)
+    def f(n):
+      time.sleep(n)
+
+    start_time = time.time()
```

Comment:
Can you document these blocks? Also, set a variable equal to 0.5 instead of referencing it by value.